### PR TITLE
util: add GrayCode

### PR DESCRIPTION
**Pull request by ekiwi**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?

#### Type of Improvement

- new feature/API

#### API Impact

- adds two new utility functions to convert binary to gray code and back

#### Backend Code Generation Impact

- this change does not impact existing code

#### Desired Merge Strategy
- squash

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Review by aswaterman — COMMENTED** [review]
This is a good util to have.

I'm pretty sure that the GrayToBinary is just something like

    Cat((in.getWidth - 2 to 0 by -1).scanLeft(in.head(1)) { case (prev, ii) => prev ^ in(ii) }

so you can get rid of the `var`.

File: src/main/scala/chisel3/util/GrayCode.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util
+
+import chisel3._
+
+object BinaryToGray {
+
+  /** Turns a binary number into gray code. */
+  def apply(in: UInt): UInt = in ^ (in >> 1.U)
```
Comment (jackkoenig) [inline src/main/scala/chisel3/util/GrayCode.scala:10]:
```suggestion
  def apply(in: UInt): UInt = in ^ (in >> 1)
```
This should be marginally more efficient in complication speed

**Review by aswaterman — APPROVED** [review]
lgtm

**Comment by ekiwi** [comment]
@jackkoenig  Did you want to backport this?

**Comment by jackkoenig** [comment]
> @jackkoenig Did you want to backport this?

Why not!

**Pull request merged**